### PR TITLE
tentative formalization of Vitali's lemma

### PR DESCRIPTION
**Pull request by affeldt-aist**
##### Motivation for this change

track C of issue #965 

(statements are being tested in other branches, to clean as soon as it stabilized)

##### Things done/to do

- [x] added corresponding entries in `CHANGELOG_UNRELEASED.md`

- [x] added corresponding documentation in the headers

##### Compatibility with MathComp 2.0

- [x] I added the label `TODO: HB port` to make sure someone ports this PR to
      the `hierarchy-builder` branch **or** I already opened an issue or PR (please cross reference).

##### Automatic note to reviewers

Read [this Checklist](https://github.com/math-comp/math-comp/wiki/Checklist-for-creating-and-review-PRs) and put a milestone if possible.

### CODE REVIEW

**Review by zstone1 — COMMENTED** [review]
Overall this looks great. I've posted a couple questions here, but nothing deep. And I haven't looked thoroughly at the vitali lemma proof itself. But the structure seems fairly reasonable.

File: theories/lebesgue_measure.v [hunk 1]
```diff
@@ -2106,3 +2106,366 @@ exists (B `|` C); split.
...
+- by move=> x /ltW/le_trans; exact.
+Qed.
+
+(* NB: can be generalized *)
+Lemma separated_countable (I : Type) (B : I -> R * {posnum R})
+    (D : set I) k : 0 < k ->
+  trivIset D (cball k \o B) -> countable D.
+Proof.
+move=> k0 tB; have [f fB] :
+    {f : I -> rat & forall i, (D i /\ ratr (f i) \in cball k (B i)) \/ ~ D i}.
```
[the remark this answers is not in the document]
Comment (affeldt-aist) [reply theories/lebesgue_measure.v:2212]: Much better indeed.

File: theories/lebesgue_measure.v [hunk 2]
```diff
@@ -2106,3 +2106,366 @@ exists (B `|` C); split.
 Qed.
 
 End egorov.
+
+Lemma properW (T : Type) (A B : set T) : A `<` B -> A `<=` B.
```
[the remark this answers is not in the document]
Comment (affeldt-aist) [reply theories/lebesgue_measure.v:2110]: Indeed, I was waiting a bit while this PR is a draft. I PRed here https://github.com/math-comp/analysis/pull/978 .

[the remark this answers is not in the document]
Comment (affeldt-aist) [reply theories/lebesgue_measure.v:2110]: Done as PR https://github.com/math-comp/analysis/pull/974. Thanks.

File: theories/lebesgue_measure.v [hunk 3]
```diff
@@ -2106,3 +2106,366 @@ exists (B `|` C); split.
...
+  by move=> c; exists (exist P _ (H _ c)) => s As; exact: bigcup_sup.
+exact: Zorn_set H.
+Qed.
+
+End Zorn_incl.
+
+Section closed_balls.
+Variable R : realType.
+
+Lemma ball_itv (x r : R) : (ball x r = `]x - r, x + r[%classic)%R.
```
[the remark this answers is not in the document]
Comment (affeldt-aist) [reply theories/lebesgue_measure.v:2138]: This also comes up in Vitali's theorem so I think it makes sense to at least put them in a section.
I'll do that in the next revision of this PR.

File: theories/lebesgue_measure.v [hunk 4]
```diff
@@ -2106,3 +2106,366 @@ exists (B `|` C); split.
...
+Proof.
+by apply/seteqP; split => y; rewrite /ball/= in_itv/= ltr_distlC.
+Qed.
+
+Lemma measurable_ball x e : measurable (@ball _ [pseudoMetricType R of R^o] x e).
```
[the remark this answers is not in the document]
Comment (affeldt-aist) [reply theories/lebesgue_measure.v]: In this case, the shorter the better.

[the remark this answers is not in the document]
Comment (affeldt-aist) [reply theories/lebesgue_measure.v]: I think that I observed that this is somehow mitigated in the MathComp 2 branch.

File: theories/lebesgue_measure.v [hunk 5]
```diff
@@ -2106,3 +2106,366 @@ exists (B `|` C); split.
...
+Proof. by rewrite ball_itv; exact: measurable_itv. Qed.
+
+Definition oball (c : R * {posnum R}) k := ball c.1 (k * c.2%:num).
```
[the remark this answers is not in the document]
Comment (affeldt-aist) [reply theories/lebesgue_measure.v]: For the time being, I will go with `scale_ball` on the model of `scale_poly`, possibly introducing locally a shortcut if it turns out to be too long. I would expect the `Z` suffix to be rather for lemmas.

File: theories/lebesgue_measure.v [hunk 6]
```diff
@@ -2106,3 +2106,366 @@ exists (B `|` C); split.
...
+
+(* NB: see also closed_ball *)
+Definition cball k (c : R * {posnum R}) := [set x | `|c.1 - x| <= k * c.2%:num].
```
[the remark this answers is not in the document]
Comment (affeldt-aist) [reply theories/lebesgue_measure.v]: I will try `scale_cball` for a start.

File: theories/lebesgue_measure.v (hunk 4 above)
[the remark this answers is not in the document]
Comment (CohenCyril) [reply theories/lebesgue_measure.v]: One one wants to see an abstract `realType` as a topological space, one must do `Import numFieldTopology.Exports.` in order to get the topology. This is a side effect of not using a canonical real numbers type.
This used to be necessary in order for results about real numbers to be applicable to either of `R`, `{x : R[i]| x \is GRing.real}` and `{x : \bar R | x \is fin_num}` (and more)... but this lemma alignment might be replaced by another mechanizm in the future and we could specialize all the results to a canonical type for the reals and this problem will disappear!

**Comment by affeldt-aist** [comment]
I still have to write the changelog and double-check the documentation,
but, modulo these two points, I'd like to have some light-review asap and
an approval to merge: if there are minor problems we will still have an
opportunity to fix them soon since at least two forthcoming PRs depend
on this one (and it is painful to systematically rebased PRs on PR for me...).

**Comment by affeldt-aist** [comment]
The main change with the original version is that we have introduced
a predicate `is_ball` so that family of balls can be represented by the
datatype `I -> set R`. That hasn't changed the structure of proofs, but
thanks to that we now have a function `scall_ball` from `set` to `set`
and we use `closure` to denote closed balls, which is closer to what
paper proofs look like.

File: theories/lebesgue_measure.v (hunk 3 above)
[the remark this answers is not in the document]
Comment (affeldt-aist) [reply theories/lebesgue_measure.v]: I have recorded this as issue #1084 so that we can merge this one when CI is green.

**Pull request merged**